picoquic/frames.c

```diff
@@ -904,32 +914,115 @@ uint8_t* picoquic_decode_stream_frame(picoquic_cnx_t* cnx, uint8_t* bytes, const
-            /* if the stream is not active yet, verify that it fits under
-             * the max stream id limit */
-             /* Check parity */
-            if (IS_CLIENT_STREAM_ID(stream->stream_id) == cnx->client_mode) {
-                if (stream->stream_id <= cnx->max_stream_id_bidir_remote) {
+    for (int nb_pass = 0; nb_pass < 2; nb_pass++) {
+        stream = cnx->first_stream;
+
+        if (nb_pass == 0) {
+            /* Skip to the first non visited stream */
+            while (stream && stream->stream_id <= cnx->last_visited_stream_id) {
+                stream = stream->next_stream;
+            }
+        }
+
+        while (stream) {
+            if ((cnx->maxdata_remote > cnx->data_sent && stream->sent_offset < stream->maxdata_remote &&
+                (stream->is_active ||
+                (stream->send_queue != NULL && stream->send_queue->length > stream->send_queue->offset) ||
+                (stream->fin_requested && !stream->fin_sent))) ||
+                    (stream->reset_requested && !stream->reset_sent) ||
+                (stream->stop_sending_requested && !stream->stop_sending_sent)) {
+                /* Something can be sent */
+                /* if the stream is not active yet, verify that it fits under
+                 * the max stream id limit */
+                 /* Check parity */
+                if (IS_CLIENT_STREAM_ID(stream->stream_id) == cnx->client_mode) {
+                    if (stream->stream_id <= cnx->max_stream_id_bidir_remote) {
+                        break;
+                    }
+                }
+                else {
+                    break;
+                }
+            }
+
+            stream = stream->next_stream;
+
+            if (nb_pass > 0) {
+                /* Dont do the loop twice */
+                if (stream && stream->stream_id > cnx->last_visited_stream_id) {
+                    stream = NULL;
                     break;
                 }
             }
+        };
+
+        /* Do only one pass if we found a stream */
+        if (stream != NULL) {
+            break;
+        }
+    }
+
+    return stream;
+}
+
+typedef struct st_picoquic_stream_data_buffer_argument_t {
+    uint8_t* bytes;
+    size_t byte_index;
+    size_t byte_space;
+    size_t space;
+    size_t length;
+    int is_fin;
+    int is_still_active;
+} picoquic_stream_data_buffer_argument_t;
+
+uint8_t* picoquic_provide_stream_data_buffer(void* context, size_t length, int is_fin, int is_still_active)
+{
+    picoquic_stream_data_buffer_argument_t * data_ctx = (picoquic_stream_data_buffer_argument_t*)context;
+    uint8_t * buffer = NULL;
+
+    if (length <= data_ctx->space) {
+        data_ctx->length = length;
+
+        if (is_fin) {
+            data_ctx->is_fin = 1;
+            data_ctx->bytes[0] |= 1;
+        }
+
+        data_ctx->is_still_active = is_still_active;
+
+        if (length < data_ctx->byte_space) {
+            if (length == data_ctx->byte_space - 1) {
```

Comment:
This special condition only makes sense when length cannot be encoded in a single byte.

Reply:
Yes, but then we are optimizing an exception case. The single byte is for data shorter than 64 bytes, the average packet length is over 1KB, so we are really looking at 1/16th of the cases. I would rather keep that code path simple, and exercised by the tests.